File: fs/elf.fs
\ =============================================================================
\  * Copyright (c) 2004, 2005 IBM Corporation
\  * All rights reserved. 
\  * This program and the accompanying materials 
\  * are made available under the terms of the BSD License 
\  * which accompanies this distribution, and is available at
\  * http://www.opensource.org/licenses/bsd-license.php
\  * 
\  * Contributors:
\  *     IBM Corporation - initial implementation
\ =============================================================================


\ ELF loader.

\ Author: Hartmut Penner <hpenner@de.ibm.com>

\ ELF 32 bit header

STRUCT
        /l field ehdr>e_ident
        /c field ehdr>e_class
        /c field ehdr>e_data
        /c field ehdr>e_version
        /c field ehdr>e_pad     
	/l field ehdr>e_ident_2
	/l field ehdr>e_ident_3
	/w field ehdr>e_type
	/w field ehdr>e_machine
	/l field ehdr>e_version
	/l field ehdr>e_entry
	/l field ehdr>e_phoff
	/l field ehdr>e_shoff
	/l field ehdr>e_flags
	/w field ehdr>e_ehsize
	/w field ehdr>e_phentsize
	/w field ehdr>e_phnum
	/w field ehdr>e_shentsize
	/w field ehdr>e_shnum
	/w field ehdr>e_shstrndx
END-STRUCT


\ ELF 32 bit program header

STRUCT
	/l field phdr>p_type
	/l field phdr>p_offset
	/l field phdr>p_vaddr
	/l field phdr>p_paddr
	/l field phdr>p_filesz
	/l field phdr>p_memsz
	/l field phdr>p_flags
	/l field phdr>p_align
END-STRUCT

\ ELF 64 bit header

STRUCT
        /l field ehdr64>e_ident
        /c field ehdr64>e_class
        /c field ehdr64>e_data
        /c field ehdr64>e_version
        /c field ehdr64>e_pad     
	/l field ehdr64>e_ident_2
	/l field ehdr64>e_ident_3
	/w field ehdr64>e_type
	/w field ehdr64>e_machine
	/l field ehdr64>e_version
	cell field ehdr64>e_entry
	cell field ehdr64>e_phoff
	cell field ehdr64>e_shoff
	/l field ehdr64>e_flags
	/w field ehdr64>e_ehsize
	/w field ehdr64>e_phentsize
	/w field ehdr64>e_phnum
	/w field ehdr64>e_shentsize
	/w field ehdr64>e_shnum
	/w field ehdr64>e_shstrndx
END-STRUCT


\ ELF 64 bit program header

STRUCT
        /l field phdr64>p_type
        /l field phdr64>p_flags    
	cell field phdr64>p_offset
	cell field phdr64>p_vaddr
	cell field phdr64>p_paddr
	cell field phdr64>p_filesz
	cell field phdr64>p_memsz
	cell field phdr64>p_align
END-STRUCT

: load-segment ( file-addr program-header-addr -- ) 

  ( file-addr  program-header-addr ) 
    dup >r phdr>p_vaddr l@ r@ phdr>p_memsz l@ erase 

  ( file-addr R: programm-header-addr )
    r@ phdr>p_vaddr l@ r@ phdr>p_memsz l@ dup 0= IF 2drop ELSE flushcache THEN

  ( file-addr R: programm-header-addr ) 
    r@ phdr>p_offset l@ +  r@ phdr>p_vaddr l@  r> phdr>p_filesz l@  move 
;


: load-segments ( file-addr -- )
  ( file-addr ) 
    dup dup ehdr>e_phoff l@ +	  \ Calculate program header address

  ( file-addr program-header-addr )
    over ehdr>e_phnum w@ 0 ?DO	  \ loop e_phnum times

  ( file-addr program-header-addr )  
      dup phdr>p_type l@ 1 = IF	  \ PT_LOAD ?
    
  ( file-addr program-header-addr )
        2dup load-segment THEN	  \ copy segment

  ( file-addr program-header-addr )
      over ehdr>e_phentsize w@ + LOOP  \ step to next header  

  ( file-addr program-header-addr )
      over ehdr>e_entry l@

  ( file-addr program-header-addr )
      nip nip 			  \ cleanup
;

: load-segment64 ( file-addr program-header-addr -- ) 

  ( file-addr  program-header-addr ) 
    dup >r phdr64>p_vaddr @ r@ phdr64>p_memsz @ erase 

  ( file-addr R: programm-header-addr )
    r@ phdr64>p_vaddr @ r@ phdr64>p_memsz @ dup 0= IF 2drop ELSE flushcache THEN
    
  ( file-addr R: programm-header-addr ) 
    r@ phdr64>p_offset @ +  r@ phdr64>p_vaddr @  r> phdr64>p_filesz @  move 
;


: load-segments64 ( file-addr -- entry )
  ( file-addr ) 
    dup dup ehdr64>e_phoff @ +	  \ Calculate program header address

  ( file-addr program-header-addr )
    over ehdr64>e_phnum w@ 0 ?DO	  \ loop e_phnum times

  ( file-addr program-header-addr )  
      dup phdr64>p_type l@ 1 = IF	  \ PT_LOAD ?
    
  ( file-addr program-header-addr )
        2dup load-segment64 THEN	  \ copy segment

  ( file-addr program-header-addr )
      over ehdr64>e_phentsize w@ + LOOP  \ step to next header  

  ( file-addr program-header-addr )
      over ehdr64>e_entry @
  
  ( file-addr program-header-addr entry )
      nip nip 			  \ cleanup
;

: elf-check-file ( file-addr --  1 : 32, 2 : 64, else bad  )
  ( file-addr )
  dup ehdr>e_ident l@ 7f454c46 <> ABORT" Not an ELF file" 
    
  ( file-addr )
  dup ehdr>e_data c@ 2 <> ABORT" Not a Big Endian ELF file" 

  ( file-addr )
  dup ehdr>e_type w@ 2 <> ABORT" Not an ELF executable" 

  ( file-addr )
  dup ehdr>e_machine w@ dup 14 <> swap 15 <> and ABORT" Not a PPC ELF executable" 

  ( file-addr)
  ehdr>e_class c@
;    

: load-elf32 ( file-addr -- )

  ( file-addr)  
  load-segments
;

: load-elf64 ( file-addr -- )

  ( file-addr)  
  load-segments64
;

: load-elf-file ( file-addr -- entry )

  ( file-addr )
  dup elf-check-file

  ( file-addr 1|2|x )

    CASE
	1 OF load-elf32 ENDOF
	2 OF load-elf64 ENDOF
	dup OF true ABORT" Neither 32- nor 64-bit ELF file" ENDOF
    ENDCASE
;

File: lib/libelf/elf.c
/******************************************************************************
 * Copyright (c) 2004, 2008 IBM Corporation
 * All rights reserved.
 * This program and the accompanying materials
 * are made available under the terms of the BSD License
 * which accompanies this distribution, and is available at
 * http://www.opensource.org/licenses/bsd-license.php
 *
 * Contributors:
 *     IBM Corporation - initial implementation
 *****************************************************************************/

/* this is elf.fs rewritten in C */

#include <string.h>
#include <cpu.h>
#include <libelf.h>

struct ehdr {
	unsigned int ei_ident;
	unsigned char ei_class;
	unsigned char ei_data;
	unsigned char ei_version;
	unsigned char ei_pad[9];
	unsigned short e_type;
	unsigned short e_machine;
	unsigned int e_version;
	unsigned int e_entry;
	unsigned int e_phoff;
	unsigned int e_shoff;
	unsigned int e_flags;
	unsigned short e_ehsize;
	unsigned short e_phentsize;
	unsigned short e_phnum;
	unsigned short e_shentsize;
	unsigned short e_shnum;
	unsigned short e_shstrndx;
};

struct phdr {
	unsigned int p_type;
	unsigned int p_offset;
	unsigned int p_vaddr;
	unsigned int p_paddr;
	unsigned int p_filesz;
	unsigned int p_memsz;
	unsigned int p_flags;
	unsigned int p_align;
};

struct ehdr64 {
	unsigned int ei_ident;
	unsigned char ei_class;
	unsigned char ei_data;
	unsigned char ei_version;
	unsigned char ei_pad[9];
	unsigned short e_type;
	unsigned short e_machine;
	unsigned int e_version;
	unsigned long e_entry;
	unsigned long e_phoff;
	unsigned long e_shoff;
	unsigned int e_flags;
	unsigned short e_ehsize;
	unsigned short e_phentsize;
	unsigned short e_phnum;
	unsigned short e_shentsize;
	unsigned short e_shnum;
	unsigned short e_shstrndx;
};

struct phdr64 {
	unsigned int p_type;
	unsigned int p_flags;
	unsigned long p_offset;
	unsigned long p_vaddr;
	unsigned long p_paddr;
	unsigned long p_filesz;
	unsigned long p_memsz;
	unsigned long p_align;
};

#define VOID(x) (void *)((unsigned long)x)

static void
load_segment(unsigned long *file_addr, struct phdr *phdr)
{
	unsigned long src = phdr->p_offset + (unsigned long) file_addr;
	/* copy into storage */
	memmove(VOID(phdr->p_vaddr), VOID(src), phdr->p_filesz);

	/* clear bss */
	memset(VOID(phdr->p_vaddr + phdr->p_filesz), 0,
	       phdr->p_memsz - phdr->p_filesz);

	if (phdr->p_memsz) {
		flush_cache(VOID(phdr->p_vaddr), phdr->p_memsz);
	}
}

static unsigned int
load_segments(unsigned long *file_addr)
{
	struct ehdr *ehdr = (struct ehdr *) file_addr;
	/* Calculate program header address */
	struct phdr *phdr =
	    (struct phdr *) (((unsigned char *) file_addr) + ehdr->e_phoff);
	int i;
	/* loop e_phnum times */
	for (i = 0; i <= ehdr->e_phnum; i++) {
		/* PT_LOAD ? */
		if (phdr->p_type == 1) {
			/* copy segment */
			load_segment(file_addr, phdr);
		}
		/* step to next header */
		phdr =
		    (struct phdr *) (((unsigned char *) phdr) +
				     ehdr->e_phentsize);
	}
	return ehdr->e_entry;
}

static void
load_segment64(unsigned long *file_addr, struct phdr64 *phdr64)
{
	unsigned long src = phdr64->p_offset + (unsigned long) file_addr;
	/* copy into storage */
	memmove(VOID(phdr64->p_vaddr), VOID(src), phdr64->p_filesz);

	/* clear bss */
	memset(VOID(phdr64->p_vaddr + phdr64->p_filesz), 0,
	       phdr64->p_memsz - phdr64->p_filesz);

	if (phdr64->p_memsz) {
		flush_cache(VOID(phdr64->p_vaddr), phdr64->p_memsz);
	}
}

static unsigned long
load_segments64(unsigned long *file_addr)
{
	struct ehdr64 *ehdr64 = (struct ehdr64 *) file_addr;
	/* Calculate program header address */
	struct phdr64 *phdr64 =
	    (struct phdr64 *) (((unsigned char *) file_addr) + ehdr64->e_phoff);
	int i;
	/* loop e_phnum times */
	for (i = 0; i <= ehdr64->e_phnum; i++) {
		/* PT_LOAD ? */
		if (phdr64->p_type == 1) {
			/* copy segment */
			load_segment64(file_addr, phdr64);
		}
		/* step to next header */
		phdr64 =
		    (struct phdr64 *) (((unsigned char *) phdr64) +
				       ehdr64->e_phentsize);
	}
	return ehdr64->e_entry;
}

#if __BYTE_ORDER == __BIG_ENDIAN
#define cpu_to_be32(x)  (x)
#else
#define cpu_to_be32(x)  bswap_32(x)
#endif

/**
 * elf_check_file tests if the file at file_addr is
 * a correct endian, ELF PPC executable
 * @param file_addr  pointer to the start of the ELF file
 * @return           the class (1 for 32 bit, 2 for 64 bit)
 *                   -1 if it is not an ELF file
 *                   -2 if it has the wrong endianess
 *                   -3 if it is not an ELF executable
 *                   -4 if it is not for PPC
 */
static int
elf_check_file(unsigned long *file_addr)
{
	struct ehdr *ehdr = (struct ehdr *) file_addr;
	/* check if it is an ELF image at all */
	if (cpu_to_be32(ehdr->ei_ident) != 0x7f454c46)
		return -1;

	/* endian check */
#if __BYTE_ORDER == __BIG_ENDIAN
	if (ehdr->ei_data != 2)
		/* not a big endian image */
#else
	if (ehdr->ei_data == 2)
		/* not a little endian image */
#endif
		return -2;

	/* check if it is an ELF executable */
	if (ehdr->e_type != 2)
		return -3;

	/* check if it is a PPC ELF executable */
	if (ehdr->e_machine != 0x14 && ehdr->e_machine != 0x15)
		return -4;

	return ehdr->ei_class;
}

/**
 * load_elf_file tries to load the ELF file specified in file_addr
 *
 * it first checks if the file is a PPC ELF executable and then loads
 * the segments depending if it is a 64bit or 32 bit ELF file
 *
 * @param file_addr  pointer to the start of the elf file
 * @param entry      pointer where the ELF loader will store
 *                   the entry point
 * @return           1 for a 32 bit file
 *                   2 for a 64 bit file
 *                   anything else means an error during load
 */
int
load_elf_file(unsigned long *file_addr, unsigned long *entry)
{
	int type = elf_check_file(file_addr);
	switch (type) {
	case 1:
		*entry = load_segments(file_addr);
		break;
	case 2:
		*entry = load_segments64(file_addr);
		break;
	}
	return type;
}
